src/main.cpp

```diff
@@ -919,19 +920,31 @@ bool ContextualCheckTransaction(
                 }
 
                 // SaplingNotePlaintext::decrypt() checks note commitment validity.
-                if (!SaplingNotePlaintext::decrypt(
+                auto encPlaintext = SaplingNotePlaintext::decrypt(
+                    chainparams.GetConsensus(),
+                    nHeight,
                     output.encCiphertext,
                     output.ephemeralKey,
                     outPlaintext->esk,
                     outPlaintext->pk_d,
-                    output.cmu)
-                ) {
+                    output.cmu);
+                if (!encPlaintext) {
                     return state.DoS(
                         DOS_LEVEL_BLOCK,
                         error("CheckTransaction(): coinbase output description has invalid encCiphertext"),
                         REJECT_INVALID,
                         "bad-cb-output-desc-invalid-encct");
                 }
+
+                // ZIP 212: Check that the note plaintexts use the v2 note plaintext
+                // version.
+                if (canopyActive && !(encPlaintext->get_lead_byte() == 0x02)) {
```

Comment:
I think that this should be

```
if (canopyActive != (encPlaintext->get_lead_byte() == 0x02)) {
```

as we don't want people putting version `0x02` plaintexts in coinbase transactions prior to the activation of Canopy, but we do want to force them to use version `0x02` once Canopy activates.

Reply:
we could use
```C++
if (!libzcash::plaintext_version_is_valid(
    chainparams.GetConsensus(), 
    nHeight, 
    encPlaintext->get_lead_byte())
) {
```
which should take care of all these cases.

(for convenient reference:)
```C++
inline bool plaintext_version_is_valid(const Consensus::Params& params, int height, unsigned char leadByte) {
    int canopyActivationHeight = params.vUpgrades[Consensus::UPGRADE_CANOPY].nActivationHeight;

    if (height < canopyActivationHeight && leadByte != 0x01) {
        // non-0x01 received before Canopy activation height
        return false;
    }
    if (height >= canopyActivationHeight
        && height < canopyActivationHeight + ZIP212_GRACE_PERIOD
        && leadByte != 0x01
        && leadByte != 0x02)
    {
        // non-{0x01,0x02} received after Canopy activation and before grace period has elapsed
        return false;
    }
    if (height >= canopyActivationHeight + ZIP212_GRACE_PERIOD && leadByte != 0x02) {
        // non-0x02 received past (Canopy activation height + grace period)
        return false;
    }
    return true;
};
```

Reply:
Hm, I think we want to compel miners to switch to the new plaintext version (aka no grace period) for these plaintexts. This change would give them a grace period that I think we don't want to give them.